### PR TITLE
[SPARK-35181][CORE] Use zstd for spark.io.compression.codec by default

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to use `zstd` as `spark.io.compression.codec` instead of `lz4` in order to reduce the disk IOs and traffic during shuffle processing and  worker decommission storage migration (between executors and to external storage).

- Since SPARK-29434 and SPARK-29576, Apache Spark 3.0+ uses ZSTD `spark.shuffle.mapStatus.compression.codec` by default instead of `GZIP`.
- Since SPARK-34503, Apache Spark 3.2 uses ZSTD for `spark.eventLog.compression.codec` by default instead of `LZ4`.

### Why are the changes needed?

To reduce the disk footprint. For TPCDS 3TB case, `zstd` has 44% less shuffle write size and 43% less shuffle read size
For some cases, the query execution with `zstd` io is 20% faster than `lz4` io.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CIs.

### CODE REVIEW

**Review by MaxGekk — COMMENTED** [review]
> to reduce the disk IOs and traffic during shuffle

@dongjoon-hyun How about CPU load. If user's job is CPU bound than zstd can introduce perf regression in the case if zstd consumes more CPU.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -52,6 +52,8 @@ class AdaptiveQueryExecSuite
 
   import testImplicits._
 
+  override protected def sparkConf = super.sparkConf.set("spark.io.compression.codec", "lz4")
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:55]: Should this be define more like a method with braces, etc or is that how similar code does it?

**Review by srowen — COMMENTED** [review]
Seems OK for consistency and maybe better speed; we should mention it in a migration guide.

**Comment by viirya** [comment]
retest this please

**Comment by LucaCanali** [comment]
I have tested this with a few runs of TPCDS query q64 that is shuffle-intensive. I see indeed a very inportant reduction of shuffle write and read bytes (about -40% as reported by  @dongjoon-hyun). Also shuffleWriteTime and shuffleFetchWaitTime are improved (about -25% and a -75%) in my test. 
I also measured an increase of CPU usage though, of about 10%. Overall the query execution time in my setup/test was basically unchanged, considering the test measurement noise. 
It still seems worth doing this, because of the large improvement on the shuffle footprint, however the increase of CPU usage should be noted as pointed out by @MaxGekk

**Review by srowen — COMMENTED** [review]
I think this is probably good idea, to trade a little more CPU for less I/O and storage. @dongjoon-hyun I'm a little concerned about changing a fairly fundamental default. Would you put this in 3.2.0? not out of the question IMHO just needs to be in a migration guide I think, and maybe worth one more ping to dev@ to solicit input.

**Comment by dongjoon-hyun** [comment]
Thank you for your feedbacks, @LucaCanali and @srowen .

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:55]: Sorry for late reply. 

For this one, we can use this form simply like the following.
```
$ git grep 'override protected def sparkConf' | grep super
sql/core/src/test/scala/org/apache/spark/sql/AggregateHashMapSuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf
sql/core/src/test/scala/org/apache/spark/sql/AggregateHashMapSuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf
sql/core/src/test/scala/org/apache/spark/sql/AggregateHashMapSuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf
sql/core/src/test/scala/org/apache/spark/sql/connector/FileDataSourceV2FallBackSuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf.set(SQLConf.USE_V1_SOURCE_LIST, "")
sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf
sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf
sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf
sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:  override protected def sparkConf = super.sparkConf.set("spark.io.compression.codec", "lz4")
sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/AlterTableRecoverPartitionsSuite.scala:  override protected def sparkConf = super.sparkConf
sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/AlterTableRecoverPartitionsSuite.scala:  override protected def sparkConf = super.sparkConf
sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcPartitionDiscoverySuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf.set(SQLConf.USE_V1_SOURCE_LIST, "")
sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamSuite.scala:  override protected def sparkConf: SparkConf = super.sparkConf
```

Or, like the following~
```
  override protected def sparkConf: SparkConf =
    super
      .sparkConf
      .set(SQLConf.USE_V1_SOURCE_LIST, "parquet")
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/CoalesceShufflePartitionsSuite.scala [hunk 2]
```diff
@@ -66,6 +66,7 @@ class CoalesceShufflePartitionsSuite extends SparkFunSuite with BeforeAndAfterAl
         .setAppName("test")
         .set(UI_ENABLED, false)
         .set(IO_ENCRYPTION_ENABLED, enableIOEncryption)
+        .set(IO_COMPRESSION_CODEC.key, "lz4")
```
Comment (mridulm) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/CoalesceShufflePartitionsSuite.scala:69]: Curious, does this test require it to be lz4 ?
Same for `AdaptiveQueryExecSuite` below - why not rely on the new default value of zstd ?

**Review by mridulm — COMMENTED** [review]
Just some minor queries, I like the change.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @mridulm . The change is required because the UT depends on the results based on the intermediate statistics of the query.

**Comment by mridulm** [comment]
I am guessing this is use of `spark.sql.adaptive.advisoryPartitionSizeInBytes` ?
Sounds good to continue using lz4 to preserve current behavior.
We can always modify the test in a later PR to adapt `targetPostShuffleInputSize` for the new codec.

**Comment by dongjoon-hyun** [comment]
Thank you. Sure, @mridulm .

**Comment by mridulm** [comment]
Looks good to me (pending other reviews comments ofcourse).
Why is this still draft btw ? Are we still testing this or waiting for other feedback/eval ?

**Comment by dongjoon-hyun** [comment]
Thank you for your comments, @mridulm . I'm looking at the stability of GitHub Action.
As you know, recently, ZStandard 1.5.0 landed at `master` branch and it seems to increase the memory usage and causes GitHub Action UT failures.

**Pull request closed without merging**